src/Operations.jl

```diff
@@ -816,28 +816,132 @@ function collect_artifacts(pkg_root::String; platform::AbstractPlatform=HostPlat
     return artifacts_tomls
 end
 
-function download_artifacts(env::EnvCache;
+function download_artifacts(ctx::Context;
                             platform::AbstractPlatform=HostPlatform(),
                             julia_version = VERSION,
-                            verbose::Bool=false,
-                            io::IO=stderr_f())
+                            verbose::Bool=false)
+    env = ctx.env
+    io = ctx.io
+    fancyprint = can_fancyprint(io)
     pkg_roots = String[]
     for (uuid, pkg) in env.manifest
         pkg = manifest_info(env.manifest, uuid)
         pkg_root = source_path(env.manifest_file, pkg, julia_version)
         pkg_root === nothing || push!(pkg_roots, pkg_root)
     end
     push!(pkg_roots, dirname(env.project_file))
+    used_artifact_tomls = Set{String}()
+    download_jobs = Channel{Function}(Inf)
+
+    print_lock = Base.ReentrantLock() # for non-fancyprint printing
+    download_states = Dict{String, Tuple{Bool,MiniProgressBar}}()
+    errors = Channel{Any}()
+    is_done = false
+    ansi_moveup(n::Int) = string("\e[", n, "A")
```

Comment:
Ideally this ANSI stuff could be moved to the progress bar code where there is something like:

```
struct MultiProgressBar
    progress_bars::Vector{MiniProgressBar}
end
```

which takes care of the printing?

Reply:
Yeah. Will handle this in a follow-on. Added a TODO